Firestore/Source/Util/FSTDispatchQueue.m

```diff
@@ -58,7 +58,10 @@ - (void)dispatchAsyncAllowingSameQueue:(void (^)(void))block {
 
 - (void)dispatchAfterDelay:(NSTimeInterval)delay block:(void (^)(void))block {
   dispatch_time_t delayNs = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delay * NSEC_PER_SEC));
-  dispatch_after(delayNs, self.queue, block);
+  dispatch_after(delayNs, self.queue, ^() {
+    // Make sure that we prioritize tasks that are already queued for immediate execution.
```

Comment:
I don't entirely understand why this double dispatch actually fixes the problem you're describing. There's still a race condition even after this, but it's now just much smaller.

`dispatch_after` still executes on the same serial queue so this block cannot be called concurrently with any other executing block. It doesn't really matter which block executes the close, does it? We still need to deal with the fact after introducing idle timeouts there are now multiple paths to closure and so the `handleStreamClose` [assertion](https://github.com/firebase/firebase-ios-sdk/blob/df76451b3101c47865ccd5721c96a37ca8c8fa70/Firestore/Source/Remote/FSTStream.m#L519) can't be correct.

In short: either the server or the client can close the connection and the event handler that closes it could be either `handleStreamClose` or `handleIdleCloseTimer`. The latter already guards against the possibility that something already closed the stream--we just need to do something similar in `handleStreamClose`.

We need to allow for the possibility that between the point where we get a callback from gRPC indicating close and the point where our own dispatch queue actually handles the server close event `handleIdleCloseTimer` could have run.

Reply:
Hm, you convinced me - there might be a slim possibility if the idle callback runs while another operation queues a depended block on the queue. So yes, it is better to just let `handleStreamClose` deal with this directly. I originally wanted to keep the state machine cleaner and avoid this extra check.